File: 01/solutionb.js
var fs = require("fs"); // imports fs

let input;
let leftArray = [];
let rightArray = [];
let total = 0;

function Solution()
{
  input = fs.readFileSync("01/sample-input.txt", "utf-8")
    .replace(/\r/gm, "")
    .split("\n")
    .map(i => i.replace(/\s+/gm, ","))
    .map(i => i.split(","));

  // separate out into two arrays
  for (let i = 0; i < input.length; i++)
  {
    leftArray.push(parseInt(input[i][0]));
    let rightValue = parseInt(input[i][1]);
    let rightIndex = rightArray.indexOf(rightValue);
    if (rightArray.some(i => i[0] === rightValue))
    {
      console.log(rightArray.indexOf(rightValue));
      rightArray[rightArray.indexOf(rightValue)][1]++;
    }
    else
    {
      rightArray.push([rightValue, 1]);
    }
  }

  leftArray.sort((a, b) => a - b);
  rightArray.sort((a, b) => a[0] - b[0]);

  console.log(rightArray);

  console.log(leftArray);
}

Solution();

File: 01/solutionb.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
	"time"
)

func main() {
	start := time.Now()
	// 1. Get input into two separate arrays
	content, err := os.ReadFile("01/input.txt")
	if err != nil {
		fmt.Printf("Failed to read file: %s\n", err)
		return
	}

	input := string(content)
	input = strings.ReplaceAll(input, "\r", "")
	lines := strings.Split(input, "\n")
	spacedRegex := regexp.MustCompile(`\s+`)
	for i, line := range lines {
		lines[i] = spacedRegex.ReplaceAllString(line, ",")
	}
	var result [][]string
	for _, line := range lines {
		result = append(result, strings.Split(line, ","))
	}
	var leftArray []int
	var rightArray []int

	for _, subArray := range result {
		leftValue, _ := strconv.Atoi(subArray[0])
		rightValue, _ := strconv.Atoi(subArray[1])

		leftArray = append(leftArray, leftValue)
		rightArray = append(rightArray, rightValue)
	}

	rightArrayMap := make(map[int]int)

	for _, v := range rightArray {
		_, found := rightArrayMap[v]
		if found {
			rightArrayMap[v]++
		} else {
			rightArrayMap[v] = 1
		}
	}

	var similarityScore int

	for _, val := range leftArray {
		frequency, found := rightArrayMap[val]
		if found {
			similarityScore += val * frequency
		}
	}

	fmt.Println(similarityScore)
	elapsed := time.Since(start)
	fmt.Printf("Elapsed time: %s\n", elapsed)
}

// execute with "go run 01/solutionb.go"
